spinalcordtoolbox/labels.py

```diff
@@ -224,52 +227,46 @@ def label_regions_from_reference(img: Image, ref: Image, centerline: bool = Fals
+
+    # Prepare to compute, for several points, the centerline index which is closest in physical space.
+    centerline_tree = KDTree(centerline_phys)
+
+    # The label coordinates.
+    labels_pix = np.argwhere(ref.data)  # shape (L, 3), int
+    labels_phys = ref.transfo_pix2phys(labels_pix, mode='absolute')  # shape (L, 3), float
+    # The corresponding label values.
+    labels_values = ref.data[tuple(labels_pix.T)]  # shape (L,), arbitrary dtype
+    # The closest centerline index for each label voxel.
+    _, labels_indices = centerline_tree.query(labels_phys)  # shape (L,), int
+
+    # Sort everything by increasing centerline index
+    sorter = np.argsort(labels_indices)
+    labels_pix = labels_pix[sorter]
+    labels_phys = labels_phys[sorter]
+    labels_values = labels_values[sorter]
+    labels_indices = labels_indices[sorter]
+    del sorter
+
+    # Add a label for above the top-most label
+    labels_values = list(labels_values)
+    labels_values.append(labels_values[-1] - 1)
+    labels_values = np.array(labels_values)
+
+    # The region we want to label.
     if centerline:
-        coordinates_input = np.concatenate((centerline_arr, np.ones((centerline_arr.shape[0], 1))), axis=1)
+        region_pix = centerline_pix
     else:
-        coordinates_input = img.getNonZeroCoordinates()
-
-    # Extract reference coordinates
-    coordinates_ref = ref.getNonZeroCoordinates(sorting='value')
-
-    # for all points in the segmentation project on the centerline and compare the `z` coordinate to the appropriate vertebral level
-    for x, y, z, _ in coordinates_input:
-        projection = project_point_on_line(point=np.array([x, y, z]), line=centerline_arr)
-        _, _, z_proj = np.rint(projection).astype(int)
-        # case 1: `z` is above the top-most disc label
-        if z_proj > coordinates_ref[0].z:
-            out.data[int(x), int(y), int(z)] = coordinates_ref[0].value - 1
-        # case 2: `z` is at or below the bottom-most disc label
-        elif z_proj <= coordinates_ref[-1].z:
-            out.data[int(x), int(y), int(z)] = coordinates_ref[-1].value
-        # case 3: `z` is between two disc labels, so find the correct vertebral level
-        else:
-            for j in range(len(coordinates_ref) - 1):
-                if coordinates_ref[j + 1].z < z_proj <= coordinates_ref[j].z:
-                    out.data[int(x), int(y), int(z)] = coordinates_ref[j].value
-
-    # Set back the original orientation
-    if img.orientation != og_img_orientation:
-        img.change_orientation(og_img_orientation)
-        out.change_orientation(og_img_orientation)
+        region_pix = np.argwhere(img.data)  # shape (R, 3), int
+    region_phys = img.transfo_pix2phys(region_pix, mode='absolute')  # shape (R, 3), float
+    # The closest centerline index for each region voxel.
+    _, region_indices = centerline_tree.query(region_phys)
```

Comment:
Actually, one more question: Is there any risk of a `KDTree` matching a segmentation point to a centerline point that's not on the same z-axis? In other words, is the projection guaranteed to always have "flat" interfaces between regions?